mode/jsx/jsx.js

```diff
@@ -0,0 +1,217 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
+// Distributed under an MIT license: http://codemirror.net/LICENSE
+
+(function(mod) {
+  if (typeof exports == "object" && typeof module == "object") // CommonJS
+    mod(require("../../lib/codemirror"), require("../javascript/javascript"),
+        require("../xml/xml"));
+  else if (typeof define == "function" && define.amd) // AMD
+    define(["../../lib/codemirror", "../javascript/javascript", "../xml/xml"],
+           mod);
+  else // Plain browser env
+    mod(CodeMirror);
+})(function(CodeMirror) {
+  "use strict";
+
+  CodeMirror.defineMode("jsx", function(config, parserConfig) {
+    var tagOpen = /^\s*</,
+        closingTagOpen = /^\s*<\//,
+        modes = {
+          'js': CodeMirror.getMode(config, {
+            name: "javascript"
+          }),
+          'xml': CodeMirror.getMode(config, {
+            name: "xml",
+            htmlMode: true,
+            multilineTagIndentFactor: parserConfig.multilineTagIndentFactor,
+            multilineTagIndentPastTag: parserConfig.multilineTagIndentPastTag,
+            allowMissing: true
+          })
+        };
+
+    function tokenBase(stream, state) {
+      if (state.nextMode) {
+        state.currentMode = state.nextMode;
+        state.nextMode = null;
+      }
+
+      // If current mode is not XML and an XML tag opening gets detected next,
+      // switch editor mode to XML.
+      // ATTENTION: This should not happen inside an attribute JavaScript
+      // expression, since XML structures are not valid XML attributes.
+      if (state.currentMode != "xml" && !state.context.inAttrJSExpression) {
+        if (stream.match(tagOpen, false)) {
+          state.currentMode = "xml";
+        }
+      }
+
+      if (state.currentMode == "xml") {
+        // If a closing tag gets detected, switch to JavaScript mode after
+        // next ">".
+        if (stream.match(closingTagOpen, false)) {
```

Comment:
Actually it's not. It is defaulting back to JavaScript and re-evaluates if it needs to switch to XML mode (https://github.com/codemirror/CodeMirror/pull/3742/files#diff-8fe55bb9634da263c3fa302cdbefb06cR42).

I could track the dangling open tags and wait for them to close before switching back to JavaScript, but I am not sure if it is a good practice. HTML tags are valid JSX syntax (http://facebook.github.io/react/docs/jsx-in-depth.html#html-tags-vs.-react-components) and from my understanding closing tags is optional in HTML5. I could implement that but sounds like it will introduce a lot of overhead, especially in the state object.